### PR TITLE
8283849: AsyncGetCallTrace may crash JVM on guarantee

### PR DESCRIPTION
**Pull request by jbachorik**
A gist of the fix is to allow relaxed instantiation of a frame when done from a signal handler - eg. for profiling purposes. 

Currently, a frame instantiation will fail on guarantee when we happen to hit a zombie method which is still on stack. While this would indicate a serious error for the normal execution flow, in case of profiling where the executing thread can be expected at any possible method this is something which may happen and we really should not take the profiled JVM down due to it.

The behaviour defaults to checking the code blob status in the guarantee so nothing will change for the rest of the callers - just ASGCT will be affected.

<hr>
Unfortunately, I am not able to create a simple reproducer for the crash other that testing in our production where the crash is happening sporadically.
However, thanks to @parttimenerd and his [ASGCT stress test](https://github.com/parttimenerd/asgct2-tester.git) the problem can be reproduced quite reliably.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8283849](https://bugs.openjdk.java.net/browse/JDK-8283849): AsyncGetCallTrace may crash JVM on guarantee

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8061/head:pull/8061` \
`$ git checkout pull/8061`

Update a local copy of the PR: \
`$ git checkout pull/8061` \
`$ git pull https://git.openjdk.java.net/jdk pull/8061/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8061`

View PR using the GUI difftool: \
`$ git pr show -t 8061`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8061.diff">https://git.openjdk.java.net/jdk/pull/8061.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ inline frame::frame() {
 
 static int spin;
 
-inline void frame::init(intptr_t* sp, intptr_t* fp, address pc) {
+inline void frame::init(intptr_t* sp, intptr_t* fp, address pc, bool forSignalHandler) {
   assert(pauth_ptr_is_raw(pc), "cannot be signed");
   intptr_t a = intptr_t(sp);
   intptr_t b = intptr_t(fp);
```
Comment (RealLucy) [inline src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:52]: What are these declarations good for?

File: src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp [hunk 2]
```diff
@@ -55,7 +55,15 @@ inline void frame::init(intptr_t* sp, intptr_t* fp, address pc) {
   _fp = fp;
   _pc = pc;
   assert(pc != NULL, "no pc?");
-  _cb = CodeCache::find_blob(pc);
+if (!forSignalHandler) {
```
Comment (RealLucy) [inline src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:58]: Indentation?

Comment (RealLucy) [inline src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:58]: Indentation?

**Review by RealLucy — CHANGES_REQUESTED** [review]
This is not a complete review yet, just some general remarks.

While I absolutely appreciate the tedious work and time you invested, I dislike the widespread nature of your changes. This is not just your fault. Your changes only make deficiencies in the code layout really obvious. 

Did you spend a thought on consolidating e.g. the changes in `os_cpu` into a common file, potentially in `os/linux` or, even better, `os/posix`? The` cpu` tree is begging for consolidation as well. Maybe most of the init stuff can be done in `runtime/frame.*`?

While browsing through the changes, I found that in `pd_get_top_frame()`, the code is inconsistent. Some instances provide special handling for `#if COMPILER2_OR_JVMCI`, others only for `#if COMPILER2`, and sometimes there is no special handling at all. We should either have it everywhere or nowhere. Furthermore, I would prefer to see only one syntactical form, either `#if`, or `#ifdef` or `#if defined()`. Please note the subtle differences!

I added a few inline comments where the code needs some TLC.

**Comment by jbachorik** [comment]
@RealLucy 
I deliberately left out any attempts on consolidation as it might make the change affecting even more files across the whole repo. From my experience it would lower the chance of getting the fix reviewed and approved significantly so I opted for the minimum number of required changes that are preventing this particular crash.

There is a JEP (https://bugs.openjdk.java.net/browse/JDK-8284289) under way which is supposed to clean up and unify the implementations for obtaining profiling stacktraces. I would prefer having the sweeping changes being rather a part of the JEP effort then attaching them to this PR. I can create a follow up JIRA ticket for the tasks you have mentioned and link it with the JEP so your input does not get lost.

File: src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp (hunk 1 above)
Comment (jbachorik) [reply src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:52]: You mean `bool forSignalHandler`?

Comment (RealLucy) [reply src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:52]: No, I am referring to lines 51 and 52. These variables are unused as far as I can see.

**Comment by RealLucy** [comment]
> @RealLucy I deliberately left out any attempts on consolidation...
> 
> There is a JEP (https://bugs.openjdk.java.net/browse/JDK-8284289) under way which is supposed to clean up...

I understand your approach - don't mess with things that are not directly affected. On the other hand: if you have to touch the code anyway, why not improve the maintainability? Maybe the JEP author has an opinion whether the cleanup could be part of the JEP. And you might get support by other reviewers as well.

**Comment by jbachorik** [comment]
> Maybe the JEP author has an opinion whether the cleanup could be part of the JEP. And you might get support by other reviewers as well.

Let me pull @parttimenerd in so he can voice his opinion about where the cleanup should happen.

File: src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp (hunk 1 above)
Comment (jbachorik) [reply src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:52]: TBH, no idea. The seem to have been added by @theRealAph some 7 years ago ...

File: src/hotspot/cpu/ppc/frame_ppc.inline.hpp [hunk 3]
```diff
@@ -33,10 +33,18 @@
 // Inline functions for ppc64 frames:
 
 // Find codeblob and set deopt_state.
-inline void frame::find_codeblob_and_set_pc_and_deopt_state(address pc) {
+inline void frame::find_codeblob_and_set_pc_and_deopt_state(address pc, bool forSignalHandler) {
   assert(pc != NULL, "precondition: must have PC");
 
-  _cb = CodeCache::find_blob(pc);
+  if (!forSignalHandler) {
```
Comment (dholmes-ora) [inline src/hotspot/cpu/ppc/frame_ppc.inline.hpp:39]: To reduce the duplication I suggest adding a helper method to CodeCache:
```
CodeBlob* find_blob(address pc, bool for_signal_handler) {
  if (!for_signal_handler) {
    return find_blob(pc);
  } else {
    CodeBlob* tmp = find_blob_unsafe(pc);
    if (tmp != NULL && temp->is_zombie()) {
      tmp = null;
    }
    return tmp;
}
```

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by parttimenerd** [comment]
The consolidation is part of my JEP and one of the main purposes. It might be easier to incorporate large clean ups into this JEP to minimize duplicated work (and code).

File: src/hotspot/cpu/ppc/frame_ppc.inline.hpp (hunk 3 above)
Comment (jbachorik) [reply src/hotspot/cpu/ppc/frame_ppc.inline.hpp]: Done

**Comment by tstuefe** [comment]
Hi,

sorry for the late chiming in.

Could we find a better way to do this that to funnel "is_signal_handler" thru every layer? How about a thread-local state instead? E.g. "bool Thread::_unsafe_code_cache_lookup"? Or, for that matter, "bool Thread::in_agct"? Could be set and restored at the beginning resp. end of AGCT.

I dislike the name "in_signal_handler" since the fact that we are inside signal handling is incidental. IIUC this is probably much more often a thread-safety issue than a reentrance issue, right? You would have the same problem if you analyzed the code cache without getting the code cache lock, while it is concurrently modified? 

So if you want to keep the argument, how about renaming it to "unsafe", or "unsafe_access", or "unsafe_codecache_lookup"? We use "unsafe" in other places as well.

Cheers, Thomas

**Comment by jbachorik** [comment]
Hi @tstuefe ,

Thanks for the review.

> Could we find a better way to do this that to funnel "is_signal_handler" thru every layer? How about a thread-local state instead? E.g. "bool Thread::_unsafe_code_cache_lookup"? Or, for that matter, "bool Thread::in_agct"? Could be set and restored at the beginning resp. end of AGCT.

I wanted to avoid as much 'magic' as possible because usually I find such PRs as having higher chance of getting reviewed. But if using a TLS to avoid funnelling extra arg is something what would usually be done I will revisit the patch and use that instead.

> I dislike the name "in_signal_handler" since the fact that we are inside signal handling is incidental. IIUC this is probably much more often a thread-safety issue than a reentrance issue, right? You would have the same problem if you analyzed the code cache without getting the code cache lock, while it is concurrently modified?

> So if you want to keep the argument, how about renaming it to "unsafe", or "unsafe_access", or "unsafe_codecache_lookup"? We use "unsafe" in other places as well.

Interestingly enough, my first sketch of the patch was using the `unsafe_access` named argument :) Then I renamed it to look less scary. If the TLS approach won't work out and I need to pass in the extra argument I can definitely rename it `unsafe_access` if that is not going to freak out the reviewers :)

**Comment by dholmes-ora** [comment]
> But if using a TLS to avoid funnelling extra arg is something what would usually be done I will revisit the patch and use that instead.

You can't set TLS from a signal handler either. Reading library-based TLS is considered safe enough.

**Comment by tstuefe** [comment]
> > But if using a TLS to avoid funnelling extra arg is something what would usually be done I will revisit the patch and use that instead.
> 
> You can't set TLS from a signal handler either. Reading library-based TLS is considered safe enough.

No need, since you only need to adjust a member inside Thread. e.g.
`Thread::current()->set_unsafe_access_mode(true)` inside AGCT.

**Comment by parttimenerd** [comment]
Using the approach proposed by Thomas has also the advantage, that we can use to handle other special casing for ASGCT too.

File: src/hotspot/cpu/x86/frame_x86.cpp [hunk 4]
```diff
@@ -715,7 +715,7 @@ intptr_t* frame::real_fp() const {
 #ifndef PRODUCT
 // This is a generic constructor which is only used by pns() in debug.cpp.
 frame::frame(void* sp, void* fp, void* pc) {
-  init((intptr_t*)sp, (intptr_t*)fp, (address)pc);
+  init((intptr_t*)sp, (intptr_t*)fp, (address)pc, CodeCache::find_blob((address)pc));
```
Comment (parttimenerd) [inline src/hotspot/cpu/x86/frame_x86.cpp:718]: There is no overload for `init` that accepts a `CodeBlob*` as its fourth argument. If it compiles then it converted implicitly into `CodeCache::find_blob((address)pc) != NULL`. Is this intentional?

**Comment by jbachorik** [comment]
Thank you all for the reviews!

I am going to close this PR and open a new one based on the thread-local flag approach as recommended by @tstuefe

The new PR is #8549

**Pull request closed without merging**